nsq/reader.py

```diff
@@ -702,14 +709,31 @@ def process_message(self, message):
 
     def giving_up(self, message):
         """
-        Called when a message has been received where ``msg.attempts > max_tries``
+        Called when a message has been received where ``msg.attempts > max_tries``.
+        
+        Failed messages will be archived to ``$data_path/$channel_$topic/%Y%m%d-%H%M%S-%f_$sequence.failed.msg``
 
-        This is useful to subclass and override to perform a task (such as writing to disk, etc.)
+        This is useful to subclass and override to perform a custom task (such as writing to disk, etc.)
 
         :param message: the :class:`nsq.Message` received
         """
-        logger.warning('[%s] giving up on message %s after %d tries (max:%d) %r',
-                       self.name, message.id, message.attempts, self.max_tries, message.body)
+
+        self.failed_count += 1
+        
+        path = os.path.join(self.data_path or "", self.topic + '_' + self.channel)
+        if not os.path.exists(path):
+            os.makedirs(path)
+        
+        date_str = datetime.datetime.utcnow().strftime("%Y%m%d-%H%M%S-%f")
+        filename = "%s_%d.failed.msg" % (date_str, self.failed_count)
+        filename = os.path.join(path, filename)
+        
+        logging.warning('[%s] giving up on message %s after %d tries (max:%d). Archived to %s %r',
+                        self.name, message.id, message.attempts, self.max_tries, filename, message.body)
+        
+        f = open(filename, 'wb')
+        f.write(message.body + '\n')
```

Comment:
not sure we should append anything to the `message.body`, in the case of a binary message format